### PR TITLE
deluge: update to 2.0.5

### PR DESCRIPTION
**Pull request by barracuda156**
#### Description

Update of `deluge` to the current version.

There is a problem with the build on old systems, which is yet to be fixed. But let me do that somewhat later in a separate PR from updating.

###### Type(s)

- [ ] bugfix
- [x] enhancement
- [ ] security fix

###### Tested on

macOS 3.2.6 Server
Xcode 3.2.6

###### Verification 
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [ ] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL? 
- [x] checked your Portfile with `port lint --nitpick`?
- [ ] tried existing tests with `sudo port test`?
- [ ] tried a full install with `sudo port -vst install`?
- [ ] tested basic functionality of all binary files?

### CODE REVIEW

**Comment by barracuda156** [comment]
Specifically, `deluge` requires `py-cryptography` >3, so existing in Macports legacy version for OS<13 does not suffice.
The last version that does not rely on Rust seems to be 3.3.2.

I have updated it (locally) and it builds fine:
```
10:~ svacchanda$ port -v installed py310-cryptography
The following ports are currently installed:
  py310-cryptography @2.9.2_2 requested_variants='' platform='darwin 10' archs='ppc' date='2022-09-09T21:33:18+0700'
  py310-cryptography @3.3.2_0 (active) requested_variants='' platform='darwin 10' archs='ppc' date='2022-09-10T04:03:07+0700'
```
(This can be made into a standalone PR, as it, apparently, gonna solve CVE CVE-2020-36242 problem. But that is after my python ports PR is merged, since `py-cryptography` is there too.)

Unlike earlier version, `py310-cryptography` 3.3.2 builds both with `openssl` 1.1 and 3. Now, the problem: `deluge` depends on `py-openssl`, which simply uses default `openssl` (so currently `openssl3`, I assume), `py-openssl` depends on `py-cryptography`, but resulting combo does not work.

When I build `py-cryptography` with 1.1 branch, `deluge` (already built) fails on launch with undefined symbol:
```
ImportError: dlopen(/opt/local/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/cryptography/hazmat/bindings/_openssl.abi3.so, 2): Symbol not found: _EVP_PKEY_set1_tls_encodedpoint
  Referenced from: /opt/local/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/cryptography/hazmat/bindings/_openssl.abi3.so
  Expected in: dynamic lookup
```
This has to do with `openssl` version compatibility: https://github.com/h2o/picotls/issues/366

When I rebuilt `py-cryptography` with 3 branch, `deluge` failed on launch with:
```
from cryptography.hazmat.bindings._openssl import ffi, lib
ImportError: dlopen(/opt/local/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/cryptography/hazmat/bindings/_openssl.abi3.so, 2): Symbol not found: _FIPS_mode_set
```
`FIPS` is something that existed in `openssl` 1.0.x.

From documentation it seems that `deluge` supports `[OpenSSL] _>= 1.0.1_`, but portfile does not have direct dependency on `openssl`, but via `py-openssl`, which, in turn, has no options to choose `openssl` branch.

@catap @dbevans @stromnov @mascguy Any ideas how to sort out this mess? :)

**Comment by catap** [comment]
As far I see:

https://github.com/pyca/cryptography/blob/82b6ce28389f0a317bc55ba2091a74b346db7cae/src/_cffi_src/openssl/fips.py#L20-L28

It expects that FIPS is always available in OpenSSL. Perhaps  some more condition may help you to exclude requirements of FIPS at `1.1` and `3.0`, anyway, I have no idea how much efforts it requires to add support of OpenSSL 3.0 to this version.

You may try to backport https://github.com/pyca/cryptography/pull/5250 to your version, but I have no idea what side effects it may introduce.

Thus, it is cryptography and any unexpected side effect may introduce security risk. So, I suggest to do it with cautions :)

**Comment by barracuda156** [comment]
@catap Wonder why existing version 2.9.x has had OpenSSL 1.1 branch set. Never tried to actually build it? :)

I will try with 1.0 and see if that helps.

**Comment by catap** [comment]
@barracuda156 to make things clear. I've started to read PR with migration to OpenSSL 3.0 and was supriside by this code: https://github.com/pyca/cryptography/pull/5250/files#diff-f5134bf8f3cf0a5cc8601df55e50697acc866c603a38caff98802bd8e17976c5R2551-R2555

It seems like a strong indicator that it uses very deep, and very internal API. It was a cause why I suggest to not try to migrate it.

But I have a kind of professional paranoia for everything that is related to security :)

**Comment by barracuda156** [comment]
> @barracuda156 to make things clear. I've started to read PR with migration to OpenSSL 3.0 and was supriside by this code: https://github.com/pyca/cryptography/pull/5250/files#diff-f5134bf8f3cf0a5cc8601df55e50697acc866c603a38caff98802bd8e17976c5R2551-R2555
> 
> It seems like a strong indicator that it uses very deep, and very internal API. It was a cause why I suggest to not try to migrate it.
> 
> But I have a kind of professional paranoia for everything that is related to security :)

@catap Thank you! That does look odd indeed :)

I had no specific intention to migrate to `openssl3` for legacy version of `py-cryptography`. I am just looking for a way to make `deluge` work.

**Comment by barracuda156** [comment]
@mascguy Let us perhaps update this first and leave fixes for old OS for a later time. `libtorrent-rasterbar` has to be fixed for Leopard to begin with, then that issue with `py-cryptography`. Alternatively, we may leave this pending, but I may not return to this soon: it appears rather complex to deal with, while at the same time there are issues to fix with a higher priority (and arguably more useful to others) – like Boost, Ruby and a bunch of smaller compilers for Haskell and Lisp.

**Comment by mascguy** [comment]
> @mascguy Let us perhaps update this first and leave fixes for old OS for a later time. `libtorrent-rasterbar` has to be fixed for Leopard to begin with, then that issue with `py-cryptography`. Alternatively, we may leave this pending, but I may not return to this soon: it appears rather complex to deal with, while at the same time there are issues to fix with a higher priority (and arguably more useful to others) – like Boost, Ruby and a bunch of smaller compilers for Haskell and Lisp.

Sounds good, let's do it!

**Comment by barracuda156** [comment]
@mascguy Yeah, I reckon there is no need to postpone this until possibly we manage to fix it for PPC. Once the latter is done, a separate PR can be made, whenever it happens.

**Comment by mascguy** [comment]
This is failing to build, due to dependencies using Python 3.11 by default.

Can you add Python 3.11 support for this port too, and make it the default?

**Comment by barracuda156** [comment]
Yes, sure. Will do now.

File: net/deluge/Portfile [hunk 1]
```diff
@@ -138,8 +114,30 @@ variant python310 conflicts python27 python39 description {Build for Python 3.10
...
+                    port:py310-asn1 \
+                    port:py310-rencode \
+                    port:py310-xdg \
+                    port:py310-openssl \
+                    port:py310-Pillow \
+                    port:py310-mako \
+                    port:py310-chardet \
+                    port:py310-six \
+                    port:py310-setproctitle \
+                    port:py310-zopeinterface
```
Comment (mascguy) [inline net/deluge/Portfile:136]: It looks like you copy-pasted the dependencies from the python310 variant, without updating.

Speaking of copy-pasta, all of the python-related logic could be simplified, by encapsulating into a proc. This is what I do for `glade`:

```
proc py_setup {python_branch} {
    global prefix frameworks_dir
    set python_version          [string map {"." ""} ${python_branch}]
    set python_framework        ${frameworks_dir}/Python.framework/Versions/${python_branch}

    configure.args-replace      -Dpython=disabled -Dpython=enabled
    configure.python            ${prefix}/bin/python${python_branch}
    configure.env-append        "PYTHON_LIBS=-L${python_framework}/lib -lpython${python_branch}" \
                                PYTHON_LIB_LOC=${python_framework}/lib
    depends_lib-append          port:py${python_version}-gobject3
    configure.pkg_config_path   ${python_framework}/lib/pkgconfig
}

variant python38 conflicts python39 python310 description {Build Python 3.8 widgets support} {
    py_setup 3.8
}

variant python39 conflicts python38 python310 description {Build Python 3.9 widgets support} {
    py_setup 3.9
}

variant python310 conflicts python38 python39 description {Build Python 3.10 widgets support} {
    py_setup 3.10
}
```

Comment (barracuda156) [reply net/deluge/Portfile:136]: I have no eyes -_-

Gonna fix this now, using your suggestion.

Comment (barracuda156) [reply net/deluge/Portfile]: Okay, another go. See if this works out.

**Comment by barracuda156** [comment]
@catap Once this is fixed and merged, could you please help me with `py-cryptography`? `libtorrent-rasterbar` gonna work on PPC once https://github.com/macports/macports-ports/pull/18810 is merged.

**Comment by barracuda156** [comment]
@mascguy Okay this still fails on dependencies, and the log is unhelpful:
```
/Users/runner/work/_temp/795bc027-bbc7-44e1-a792-eaddb57dcaf1.sh: line 77:  3263 Terminated: 15          tail -f "$workdir/logs/dependencies-progress.txt" 2> /dev/null
```
Likely I did something wrong, but could you take a look, please?

**Comment by mascguy** [comment]
> @mascguy Okay this still fails on dependencies, and the log is unhelpful:
> 
> ```
> /Users/runner/work/_temp/795bc027-bbc7-44e1-a792-eaddb57dcaf1.sh: line 77:  3263 Terminated: 15          tail -f "$workdir/logs/dependencies-progress.txt" 2> /dev/null
> ```

It looks like `py-rencode` didn't have a python 3.11 subport, but took care of that in the latest push.

**Review by mascguy — APPROVED** [review]
LGTM, assuming CI passes

**Comment by mascguy** [comment]
Ci was failing for `py310-rencode`, so committed that directly to master and rebased. Let's see if it works this time.

**Comment by barracuda156** [comment]
@mascguy Thanks! Once this is sorted, it would be actually easier to work on PPC fix separately. Otherwise it becomes a mess with all these python deps.

**Comment by mascguy** [comment]
Testing locally, I'm seeing the following, as `libtorrent-rasterbar` needs a rev-bump:

```
$ sudo port -v rev-upgrade
--->  Scanning binaries for linking errors
Could not open /opt/local/libexec/boost/1.76/lib/libboost_python311-mt.dylib:
Error opening or reading file (referenced from /opt/local/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/libtorrent.cpython-311-darwin.so)
--->  Found 1 broken file, matching files to ports       
--->  Found 1 broken port, determining rebuild order
You can always run 'port rev-upgrade' again to fix errors.
The following ports will be rebuilt: libtorrent-rasterbar @2.0.9+python311
```

Just took care of that via master, and I'll rebase this PR momentarily.

**Comment by mascguy** [comment]
Even with `libtorrent-rasterbar` fixed, I'm still seeing the following issue. (And I think I saw the same thing in the previous CI build.)

```
Traceback (most recent call last):
  File "/opt/local/var/macports/build/_opt_macports_sources_ports_latest_net_deluge/deluge/work/deluge-2.0.5/deluge/_libtorrent.py", line 24, in <module>
    import deluge.libtorrent as lt
ModuleNotFoundError: No module named 'deluge.libtorrent'
```

And based on some quick Googling, it appears that we may be missing a component. (Such as the Python bindings for `libtorrent`.)

Just to confirm, this is successfully building for you locally? If so, is there anything you forgot to add to this PR...?

**Comment by mascguy** [comment]
> Even with `libtorrent-rasterbar` fixed, I'm still seeing the following issue. (And I think I saw the same thing in the previous CI build.)
> 
> ```
> Traceback (most recent call last):
>   File "/opt/local/var/macports/build/_opt_macports_sources_ports_latest_net_deluge/deluge/work/deluge-2.0.5/deluge/_libtorrent.py", line 24, in <module>
>     import deluge.libtorrent as lt
> ModuleNotFoundError: No module named 'deluge.libtorrent'
>
> [...stack traces snipped...]
>
> SystemError: type Boost.Python.enum has the Py_TPFLAGS_HAVE_GC flag but has no traverse function
> ```

Okay, after doing some research, this appears to be a Python 3.11-specific issue. And it relates to stricter checks, for enums:

https://github.com/python/cpython/issues/88429

But ultimately the issue lies with our Boost ports, which need the following upstream fix for Python 3.11. (CC: @michaelld)

https://github.com/boostorg/python/commit/a218babc8daee904a83f550fb66e5cb3f1cb3013

**Comment by mascguy** [comment]
In the interim, make Python 3.10 the default, and we'll tackle the Boost fix separately. (It'll potentially need to be applied to all Boost versions, so it's not a 5-minute change.)

CI will still fail, but looks good locally... so I'll merge once you've made that update.

**Comment by barracuda156** [comment]
@mascguy So we pause with this one for a while?

**Comment by mascguy** [comment]
> @mascguy So we pause with this one for a while?

Nope, see my later comment, re: making Python 3.10 the default.  :-)

**Review by mascguy — APPROVED** [review]

**Comment by barracuda156** [comment]
> > @mascguy So we pause with this one for a while?
> 
> Nope, see my later comment, re: making Python 3.10 the default. :-)

Ah, okay. Good.

**Pull request merged**